### DISCUSSION TITLE
tm1py-samples: dim_create.py with attribute values

### DISCUSSION THREAD

**Discussion by ubihw** [cubewise-code/tm1py#781] [General]
Hello,

https://github.com/cubewise-code/tm1py-samples/blob/master/Metadata/dim_create.py shows how to create a dimension, including attributes. What is missing, is, how the attribute values are written. I noticed, that the Element Object accepts an attributes list as additional parameter, and extended the example like follows:

```
from TM1py.Objects import Dimension, Element, ElementAttribute, Hierarchy
from TM1py.Services import TM1Service

from config import tm1params

name = 'TM1py Region'

# Connection to TM1. Needs IP, Port, Credentials, and SSL
with TM1Service(**tm1params) as tm1:
    # create elements objects
    # the attribute values must be in the same order as the attributes in element_attributes!
    elements = [Element(name='Europe', element_type='Consolidated', attributes=['Europe','EUR','mixed']),
                Element(name='CH', element_type='Numeric', attributes=['Switzerland','SWI','CHF']),
                Element(name='UK', element_type='Numeric', attributes=['United Kingdom','UnK','BP']),
                Element(name='BE', element_type='Numeric', attributes=['Belgium','BEL','EUR'])]

    # create edge object
    edges = {('Europe', 'CH'): 1,
             ('Europe', 'UK'): 1,
             ('Europe', 'BE'): 1}

    # create the element_attributes
    element_attributes = [ElementAttribute(name='Name Long', attribute_type='Alias'),
                          ElementAttribute(name='Name Short', attribute_type='Alias'),
                          ElementAttribute(name='Currency', attribute_type='String')]

    # create hierarchy object
    hierarchy = Hierarchy(name=name, dimension_name=name, elements=elements, element_attributes=element_attributes,
                          edges=edges)

    # create dimension object
    d = Dimension(name=name, hierarchies=[hierarchy])

    # create dimension in TM1 !
    tm1.dimensions.create(d)

    # populate attribute cube
    attrcube = '}ElementAttributes_' + name
    attrvals = dict()
    for i in range(len(element_attributes)):
        attrvals.update({(e.name,element_attributes[i].name) : e.element_attributes[i] for e in elements})
    tm1.cubes.cells.write_values(attrcube,attrvals)
```

It worked for me, but I dont know, if I used the attributes parameter in the way it was intended for. If there is a more TM1py-ish way to populate the attributes, please let me know. And, maybe dim_create.py could be extended with the right way to populate attributes?

**Comment by MariusWirtz** [comment]
Hi @ubihw,

TM1py treats attribute values as cell values. So to write attribute values you can write directly to the `}ElementAttributes` cube, just like you did.
In many ways, the `write_values` function is inferior to the standard `write` function.  You can read about the different write functions here:
https://cubewise-code.github.io/tm1py-tales/2021/write-with-tm1py.html
I think the only real use case of the `write_values` function is to use data spread shortcuts on cells. 

Regarding the attributes parameter in the `Element` constructor.
One could assume that the attribute values are updated in TM1 after this line is executed.

```
tm1.dimensions.create(d)
```

Unfortunately, TM1py can't send the attribute values in one operation together with the element updates to TM1. 
So the `attributes` parameter in the Element constructor is only used when TM1py retrieves data from TM1 hierarchies. For write-back you need to write straight to the attributes cube, just like you did.

Yes. I think the way you use the attributes property to store the attribute values in the elements makes sense. 

You can read more about dimension updates here:
https://cubewise-code.github.io/tm1py-tales/2021/sync-dimensions.html
